Installation issue: kokkos-kernels@3.4.01 with rocm

### Steps to reproduce the issue

```console
spack install kokkos-kernels ^kokkos+rocm amdgpu_target=gfx908
```

### Information on your system

```console

$ spack debug report
* **Spack:** 0.17.0-316-8ff81d4
* **Python:** 3.6.8
* **Platform:** linux-centos7-broadwell
* **Concretizer:** clingo
```

### Additional information

```
g++: error: unrecognized command-line option '-fno-gpu-rdc'
g++: error: unrecognized command-line option '--amdgpu-target=gfx908'
g++: error: unrecognized command-line option '--amdgpu-target=gfx908'
make[2]: *** [src/CMakeFiles/kokkoskernels.dir/impl/generated_specializations_cpp/abs/Blas1_abs_eti_DOUBLE_LAYOUTLEFT_EXECSPACE_SERIAL_MEMSPACE_HOSTSPACE.cpp.o] Error 1

```

[spack-build-out.txt](https://github.com/spack/spack/files/7615117/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/7615118/spack-build-env.txt)

cc: @brian-kelley  @lucbv  @srajama1

BTW: Should kokkos-kernels support a `rocm` variant?

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/kokkos-kernels/package.py
+++ b/var/spack/repos/builtin/packages/kokkos-kernels/package.py
@@ -102,8 +102,11 @@ class KokkosKernels(CMakePackage, CudaPackage):
             options.append("-DSpack_WORKAROUND=On")
 
         options.append("-DKokkos_ROOT=%s" % spec["kokkos"].prefix)
-        # Compiler weirdness due to nvcc_wrapper
-        options.append("-DCMAKE_CXX_COMPILER=%s" % spec["kokkos"].kokkos_cxx)
+        if spec.satisfies('^kokkos+rocm'):
+            options.append("-DCMAKE_CXX_COMPILER=%s" % spec['hip'].hipcc)
+        else:
+            # Compiler weirdness due to nvcc_wrapper
+            options.append("-DCMAKE_CXX_COMPILER=%s" % spec["kokkos"].kokkos_cxx)
 
         if self.run_tests:
             options.append("-DKokkosKernels_ENABLE_TESTS=ON")

--- a/var/spack/repos/builtin/packages/petsc/package.py
+++ b/var/spack/repos/builtin/packages/petsc/package.py
@@ -220,6 +220,12 @@ class Petsc(Package, CudaPackage, ROCmPackage):
     depends_on('hip', when='+rocm')
     depends_on('hipblas', when='+rocm')
     depends_on('hipsparse', when='+rocm')
+    depends_on('rocsparse', when='+rocm')
+    depends_on('rocsolver', when='+rocm')
+    depends_on('rocblas', when='+rocm')
+    depends_on('rocrand', when='+rocm')
+    depends_on('rocthrust', when='+rocm')
+    depends_on('rocprim', when='+rocm')
 
     # Build dependencies
     depends_on('python@2.6:2.8', type='build', when='@:3.10')
@@ -320,6 +326,7 @@ class Petsc(Package, CudaPackage, ROCmPackage):
     depends_on('kokkos-kernels', when='+kokkos')
     depends_on('kokkos+cuda+wrapper+cuda_lambda', when='+kokkos +cuda')
     depends_on('kokkos-kernels+cuda', when='+kokkos +cuda')
+    depends_on('kokkos+rocm', when='+kokkos +rocm')
 
     # Using the following tarballs
     # * petsc-3.12 (and older) - includes docs
@@ -413,7 +420,7 @@ class Petsc(Package, CudaPackage, ROCmPackage):
         # if not (useinc || uselib): usedir - i.e (False, False)
         for library in (
                 ('cuda', 'cuda', False, False),
-                ('hip', 'hip', False, False),
+                ('hip', 'hip', True, False),
                 'metis',
                 'hypre',
                 'parmetis',
@@ -495,6 +502,22 @@ class Petsc(Package, CudaPackage, ROCmPackage):
                 else:
                     options.append('CUDAFLAGS=-gencode arch=compute_{0},code=sm_{0}'
                                    .format(cuda_arch[0]))
+        if '+rocm' in spec:
+            if not spec.satisfies('amdgpu_target=none'):
+                hip_arch = spec.variants['amdgpu_target'].value
+                options.append('--with-hip-arch={0}'.format(hip_arch[0]))
+            hip_pkgs = ['hipsparse', 'hipblas', 'rocsparse', 'rocsolver', 'rocblas']
+            hip_ipkgs = hip_pkgs + ['rocthrust', 'rocprim']
+            hip_lpkgs = hip_pkgs + ['rocrand']
+            hip_inc = ''
+            hip_lib = ''
+            for pkg in hip_ipkgs:
+                hip_inc += spec[pkg].headers.include_flags + ' '
+            for pkg in hip_lpkgs:
+                hip_lib += spec[pkg].libs.joined() + ' '
+            options.append('HIPPPFLAGS=%s' % hip_inc)
+            options.append('with-hip-lib=%s -L%s -lamdhip64' %
+                           (hip_lib, spec['hip'].prefix.lib))
 
         if 'superlu-dist' in spec:
             if spec.satisfies('@3.10.3:3.15'):

